### DISCUSSION TITLE
What is `replace` in `TrainState`

### DISCUSSION THREAD

**Discussion by pablo2909** [google/flax#2042] [Q&A] [answered]
Hello, 

I am looking at the code of `TrainState` in `flax.training.train_state.py`. The function ` apply_gradients` calls `self.replace` that is not defined in `TrainState`, but in `PyTreeNode` that only raises a `NotImplementedError`. However when we run it, it does not raise it. 
Am I missing something obvious ? I am not sure how this works.

Thank you very much

**Comment by marcvanzee** [comment, marked as answer]
`PyTreeNode` implements the [__init_subclass__](https://github.com/google/flax/blob/193f44e9f620ac493d662249741e9acc47d637cd/flax/struct.py#L198) method, which will converts the subclass (in this case `TrainState`) into a Flax dataclass, which does implement the [replace](https://github.com/google/flax/blob/193f44e9f620ac493d662249741e9acc47d637cd/flax/struct.py#L109) method.

We inherit from `PyTreeNode` to avoid type checking issues when using pytype.

**Comment by pablo2909** [comment]
I see thank you very much !
